src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs

```diff
@@ -100,25 +126,84 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
                 while (userName.Length == 0);
             }
 
-            passwordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_Password, userName
-            );
+            passwordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_Password, userName);
 
-            //
             // now, prompt for the password
-            //
-            WriteToConsole(passwordPrompt, true);
-            password = ReadLineAsSecureString();
-            if (password == null)
+            do
             {
-                return null;
+                WriteToConsole(passwordPrompt, true);
+                password = ReadLineAsSecureString();
+                if (password == null)
+                {
+                    return null;
+                }
             }
+            while (password.Length == 0);
 
-            WriteLineToConsole();
+            if (reenterPassword)
+            {
+                reenterPasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_ReenterPassword, userName);
+                passwordMismatch = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_PasswordMismatch);
 
-            cred = new PSCredential(userName, password);
+                // now, prompt to re-enter the password.
+                WriteToConsole(reenterPasswordPrompt, true);
+                confirmPassword = ReadLineAsSecureString();
+                if (confirmPassword == null)
+                {
+                    return null;
+                }
 
+                if (!SecureStringEquals(password, confirmPassword))
+                {
+                    WriteToConsole(ConsoleColor.Red, ConsoleColor.Black, passwordMismatch, false);
+                    return null;
+                }
+            }
+
+            WriteLineToConsole();
+            cred = new PSCredential(userName, password);
             return cred;
         }
+
+        private static bool SecureStringEquals(SecureString password, SecureString confirmPassword)
+        {
+
+            IntPtr pwd_ptr = IntPtr.Zero;
+            IntPtr confirmPwd_ptr = IntPtr.Zero;
+            try
+            {
+                pwd_ptr = Marshal.SecureStringToBSTR(password);
+                confirmPwd_ptr = Marshal.SecureStringToBSTR(confirmPassword);
+
+                int pwdLength = Marshal.ReadInt32(pwd_ptr, -4);
+                int confirmPwdLength = Marshal.ReadInt32(confirmPwd_ptr, -4);
+                if(pwdLength != confirmPwdLength)
+                {
+                    return false;
+                }
+
+                int equal = 0;
```

Comment:
It is not clear to me that doing multiple PInvoke reads is that performant, or that we really need to worry about performance at this level.  For readability and maintainability I am inclined to simply copy the byte array and compare bytes (zero out array in finally block).  Also note that this loop can be exited as soon as one byte mismatch is found.

Reply:
@PaulHigin Is the below snippet as per requirements. 
```
            int size = password.Length * 2;
            IntPtr pwd_ptr = IntPtr.Zero;
            IntPtr confirmPwd_ptr = IntPtr.Zero;
            byte[] pwd = new byte[size];
            byte[] confirmPwd = new byte[size];
            try
            {
                pwd_ptr = Marshal.SecureStringToBSTR(password);
                Marshal.Copy(pwd_ptr, pwd, 0, size);

                confirmPwd_ptr = Marshal.SecureStringToBSTR(confirmPassword);
                Marshal.Copy(confirmPwd_ptr, confirmPwd, 0, size);

                return pwd.SequenceEqual(confirmPwd);
            }
            finally
            {
                if (pwd_ptr != IntPtr.Zero)
                {
                    Marshal.ZeroFreeBSTR(pwd_ptr);
                }

                if (confirmPwd_ptr != IntPtr.Zero)
                {
                    Marshal.ZeroFreeBSTR(confirmPwd_ptr);
                }

                Array.Clear(pwd, 0, size);
                Array.Clear(confirmPwd, 0, size);
                size = 0;
            }
```

Reply:
The pattern with `Marshal.ReadInt32` comes from StackOverflow site and could is used broadly. It and look good for me.

Reply:
I apologize for the churn.  There are various ways to do this and I feel that perf differences are not significant.  I am fine with it either way and will leave it up to the author, however I do want to see an early out for first mismatch in the loop (which was missing before).

Reply:
@iSazonov @PaulHigin 
I have made the changes. Please review and let me know your feedback.